smetrics/src/test/scala/com/evolutiongaming/smetrics/MeasureDurationSpec.scala

```diff
@@ -42,31 +79,30 @@ object MeasureDurationSpec {
     val Empty: State = State(List.empty)
   }
 
-
-  type StateT[A] = cats.data.StateT[Id, State, A]
-
-  object StateT {
-
-    def apply[A](f: State => (State, A)): StateT[A] = cats.data.StateT[Id, State, A](f)
-  }
-
-
-  implicit val clockStateT: Clock[StateT] = new Clock[StateT] {
-
-    def realTime(unit: TimeUnit) = {
-      StateT { state =>
-        val (state1, timestamp) = state.timestamp
-        val timestamp1 = unit.convert(timestamp, TimeUnit.NANOSECONDS)
-        (state1, timestamp1)
-      }
+  implicit def timerStateT[F[_]: Applicative]: Timer[StateT[F, State, *]] = new Timer[StateT[F, State, *]] {
+
+    def clock: Clock[StateT[F, State, *]] = new Clock[StateT[F, State, *]] {
+      def realTime(unit: TimeUnit): StateT[F, State, Long] =
+        StateT { state =>
+          val (state1, timestamp) = state.timestamp
+          val timestamp1 = unit.convert(timestamp, TimeUnit.NANOSECONDS)
+          (state1, timestamp1).pure[F]
+        }
+
+      def monotonic(unit: TimeUnit): StateT[F, State, Long] =
+        StateT { state =>
+          val (state1, timestamp) = state.timestamp
+          val timestamp1 = unit.convert(timestamp, TimeUnit.NANOSECONDS)
+          (state1, timestamp1).pure[F]
+        }
     }
 
-    def monotonic(unit: TimeUnit) = {
+    def sleep(duration: FiniteDuration): StateT[F, State, Unit] =
       StateT { state =>
         val (state1, timestamp) = state.timestamp
-        val timestamp1 = unit.convert(timestamp, TimeUnit.NANOSECONDS)
-        (state1, timestamp1)
+        val timestamp1 = timestamp + duration.toNanos
+        val newState = state1.copy(timestamps = state1.timestamps :+ timestamp1)
```

Comment:
Looks like you want to add the sleep duration to all timestamps in the state, no?
Otherwise you could have time running backwards, like `State(1000L :: 2000L :: 500L :: Nil)` if you sleep for 500 nanos